### ISSUE TITLE
Possible Sinkhorn normalization over wrong axis when last dimension is singleton

### ISSUE THREAD

**Issue by arishov1** [lucidrains/hyper-connections#9]
Hi! I may be misunderstanding the intended tensor layout here, but I noticed a potential issue with sinkhorn_knopps.

sinkhorn_knopps assumes the transport matrix lives in the last two dimensions:

```
alpha = l1norm(alpha, dim=-2)
alpha = l1norm(alpha, dim=-1)
```

In my runs, the tensor passed to Sinkhorn had shape:
`torch.Size([1, 65, 1, 4, 1, 1, 4, 1])`

Here, dim=-1 has size 1, so the second normalization becomes a no-op and Sinkhorn effectively normalizes only one axis.

Is Sinkhorn expected to always receive tensors shaped as (..., streams, streams) at this point?

Thanks, and apologies if I’m missing something obvious.

**Comment by arishov1** [comment]
Happy to review the latest, thanks!
I’ll take a look and let you know.

**Comment by arishov1** [comment]
Reviewed the latest commit,  shapes look correct now, thanks for the quick fix!
Small follow-up question: have you considered log-domain Sinkhorn for extra numerical stability vs exp + L1 renorm? Not sure it’s needed, just curious if you’ve tried it.

**Comment by lucidrains** [comment]
@arishov1 no, but that's a good idea, added it to the repo!

**Comment by hjc18** [comment]
Hi @lucidrains, this is also an equivalent implementation to @arishov1's log-space iteration :)

```python
import torch.nn.functional as F
# iteration body:
log_alpha = F.log_softmax(log_alpha, dim = -2)
log_alpha = F.log_softmax(log_alpha, dim = -1)
```

**Comment by hjc18** [comment]
> [@hjc18](https://github.com/hjc18) i like that Jiechen! made the switch!

Than you!
